tests/testsuite/rename_deps.rs

```diff
@@ -145,3 +147,135 @@ fn rename_with_different_names() {
+            r#"
+                [package]
+                name = "foo"
+                version = "0.1.0"
+                authors = []
+            "#,
+        )
+        .file("src/lib.rs", "pub fn foo3() {}")
+        .build();
+
+    let p = project("foo")
+        .file(
+            "Cargo.toml",
+            &format!(r#"
+                cargo-features = ["alternative-registries", "rename-dependency"]
+
+                [package]
+                name = "test"
+                version = "0.1.0"
+                authors = []
+
+                [dependencies]
+                foo = "0.2"
+                foo1 = {{ version = "0.1", package = "foo" }}
+                foo2 = {{ version = "0.1", registry = "alternative", package = "foo" }}
+                foo3 = {{ git = '{}', package = "foo" }}
+                foo4 = {{ path = "foo", package = "foo" }}
+            "#,
+            g.url())
+        )
+        .file(
+            "src/lib.rs",
+            "
+                extern crate foo;
+                extern crate foo1;
+                extern crate foo2;
+                extern crate foo3;
+                extern crate foo4;
+
+                pub fn foo() {
+                    foo::foo();
+                    foo1::foo1();
+                    foo2::foo2();
+                    foo3::foo3();
+                    foo4::foo4();
+                }
+            ",
+        )
+        .file(
+            "foo/Cargo.toml",
+            r#"
+                [project]
+                name = "foo"
+                version = "0.1.0"
+                authors = []
+            "#,
+        )
+        .file("foo/src/lib.rs", "pub fn foo4() {}")
+        .build();
+
+    assert_that(
+        p.cargo("build -v").masquerade_as_nightly_cargo(),
+        execs().with_status(0),
+    );
+}
+
+#[test]
+fn rename_and_patch() {
+    Package::new("foo", "0.1.0").publish();
+
+    let p = project("foo")
+        .file(
+            "Cargo.toml",
+            r#"
+                cargo-features = ["rename-dependency"]
+
+                [package]
+                name = "test"
+                version = "0.1.0"
+                authors = []
+
+                [dependencies]
+                bar = { version = "0.1", package = "foo" }
+
+                [patch.crates-io]
+                foo = { path = "foo" }
+            "#,
+        )
+        .file(
+            "src/lib.rs",
+            "
+                extern crate bar;
+
+                pub fn foo() {
+                    bar::foo();
+                }
+            ",
+        )
+        .file(
+            "foo/Cargo.toml",
+            r#"
+                [project]
+                name = "foo"
+                version = "0.1.0"
+                authors = []
+            "#,
+        )
+        .file("foo/src/lib.rs", "pub fn foo() {}")
+        .build();
+
+    assert_that(
+        p.cargo("build -v").masquerade_as_nightly_cargo(),
+        execs().with_status(0),
+    );
+}
```

Comment:
In the implementation, we use `Vec<Dependency>` to store dependencies between two packages. Why do we need a `Vec` and not just a `Depedendency`? Do we allow to depend on the same crate under different names? Or is it because a dependency might be duplicated in `[dependencies]` and `[build-dependnecies]`? Is `rustc` ok with several `--extern` pointing to the same crate? Let's have a test for it? 

Ah, now I see it is handled in code by `multiple dependencies listed for the same crate` error, so let's have a test for exactly that error message!

Reply:
Speaking of tests, let's also add a test that checks that if we rename dependency in Cargo.toml, then we'll rebuild the code? Looking at the fingerprint calculation, looks like we might not do it? Perhaps this is a separate issue unrelated to the PR though :)

Reply:
>Speaking of tests, let's also add a test that checks that if we rename dependency in Cargo.toml, then we'll rebuild the code?

Added a test in https://github.com/rust-lang/cargo/pull/5425. Looks like we indeed don't handle it yet, will look into fixing that!

Reply:
Er oops sorry meant to do this and forgot, thanks though for picking it up!